File: src/PointCloud.cpp
#include "PointCloud.h"


PointCloud::PointCloud() {
    point_cloud.reset(new pcl::PointCloud<pcl::PointXYZ>());
}

PointCloud::PointCloud(pcl::PointCloud<pcl::PointXYZ>::Ptr& ptr) {
    point_cloud = ptr;
}

void PointCloud::add_point(double x, double y, double z) {
    pcl::PointXYZ o;
    o.x = x;
    o.y = y;
    o.z = z;
    point_cloud->push_back(o);
}

double PointCloud::get_x(unsigned int num) {
    return point_cloud->at(num).x;
}

double PointCloud::get_y(unsigned int num) {
    return point_cloud->at(num).y;
}

double PointCloud::get_z(unsigned int num) {
    return point_cloud->at(num).z;
}

unsigned int PointCloud::get_size() {
    return point_cloud->size();
}

PointCloud& PointCloud::statistical_outlier_removal(int mean, double dev) {
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_filtered (new pcl::PointCloud<pcl::PointXYZ>);
    pcl::StatisticalOutlierRemoval<pcl::PointXYZ> sor;
    sor.setInputCloud (point_cloud);
    sor.setMeanK (mean);
    sor.setStddevMulThresh (dev);
    sor.filter (*cloud_filtered);
    PointCloud* pc = new PointCloud(cloud_filtered);
    return *pc;
}

PointCloud& PointCloud::approximate_voxel_grid(double x, double y , double z) {
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_filtered (new pcl::PointCloud<pcl::PointXYZ>);
    pcl::ApproximateVoxelGrid<pcl::PointXYZ> voxel_filter;
    voxel_filter.setLeafSize (x, y, z);
    voxel_filter.setInputCloud (point_cloud);
    voxel_filter.filter (*cloud_filtered);
    PointCloud* pc = new PointCloud(cloud_filtered);
    return *pc;   
}

PointCloud& PointCloud::icp(int i, double d1, double d2, double d3, PointCloud& source) {
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_result (new pcl::PointCloud<pcl::PointXYZ>);
    pcl::IterativeClosestPoint<pcl::PointXYZ, pcl::PointXYZ> registration;
    registration.setInputSource(source.point_cloud);
    registration.setInputTarget(point_cloud);
    registration.setMaximumIterations(i);
    registration.setMaxCorrespondenceDistance (d1);
    registration.setEuclideanFitnessEpsilon (d2);
    registration.setTransformationEpsilon (d3);
    registration.align(*cloud_result);

    if(registration.hasConverged()) std::cout << "success" << std::endl;
    else std::cout << "fail" << std::endl;
    
    PointCloud* pc = new PointCloud(cloud_result);
    return *pc;
}

PointCloud& PointCloud::ndt(int i, double d1, double d2, double d3, PointCloud& source) {
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_result (new pcl::PointCloud<pcl::PointXYZ>);
    pcl::NormalDistributionsTransform<pcl::PointXYZ, pcl::PointXYZ> registration;
    registration.setInputSource(source.point_cloud);
    registration.setInputTarget(point_cloud);
    registration.setMaximumIterations (i);
    registration.setStepSize (d1);
    registration.setResolution (d2);
    registration.setTransformationEpsilon (d3);
    registration.align(*cloud_result);

    if(registration.hasConverged()) std::cout << "success" << std::endl;
    else std::cout << "fail" << std::endl;

    PointCloud* pc = new PointCloud(cloud_result);
    return *pc;   
}

void PointCloud::read_from_file(char* filename) {
    pcl::PCDReader reader;
    reader.read<pcl::PointXYZ> (filename, *point_cloud);
}

void PointCloud::write_to_file(char* filename) {
    pcl::PCDWriter writer;
    writer.write<pcl::PointXYZ> (filename, *point_cloud, false);
}

File: src/PointCloud.py
#! /usr/bin/python3

import PointCloudWrapper as pcw

class PointCloud:

    def __init__(self, *pc):
        if pc:
            self.pc = pc[0]
        else:
            self.pc = pcw.PointCloud()

    def size(self):
        return self.pc.get_size()

    def add(self, point):
        x, y, z = point
        self.pc.add_point(x, y, z)

    def get(self, num):
        x = self.pc.get_x(num)
        y = self.pc.get_y(num)
        z = self.pc.get_z(num)
        return (x, y, z)

    def static(self, mean=50, dev=1.):
        tmp = self.pc.statistical_outlier_removal(mean, dev)
        return PointCloud(tmp)

    def voxel(self, leaf_x=0.1, leaf_y=0.1, leaf_z=0.1):
        tmp = self.pc.approximate_voxel_grid(leaf_x, leaf_y, leaf_z)
        return PointCloud(tmp)

    def icp(self, source, iterations=1000, dist=0.05, trans=1e-8, fitness=1.):
        tmp = self.pc.icp(iterations, dist, fitness, trans, source.pc)
        return PointCloud(tmp)

    def ndt(self, source, iterations=35, step=0.01, resol=1.0, trans=0.01):
        tmp = self.pc.ndt(iterations, step, resol, trans, source.pc)
        return PointCloud(tmp)

    def write_normal(self, filename, radius=10):
        self.pc.normal(filename, radius)

    def read(self, filename):
        self.pc.read_from_file(filename)

    def concat(self, pc):
        return PointCloud(self.pc.concat(pc.pc))

    @classmethod
    def read_file(cls, filename):
        tmp = PointCloud(pcw.PointCloud())
        tmp.read(filename)
        return tmp

    def write(self, filename):
        self.pc.write_to_file(filename)

# pc1 = PointCloud.read_file("../../data/file00.pcd")
# pc2 = PointCloud.read_file("../../data/file01.pcd")

# pc1 = pc1.voxel(0.0001, 0.0001, 0.0001)
# pc2 = pc2.voxel(0.0001, 0.0001, 0.0001)
# pc1 = pc1.static(dev=0.1)
# pc2 = pc2.static(dev=0.1)
# pc1.write("../../data/target.pcd")

# pc3 = pc1.ndt(pc2, iterations=100, step=0.0001, resol=0.0005, trans=0.0001)
# pc3.write("../../data/result.pcd")

# pc.read("../../data/file01.pcd")

# pc = pc.removal(mean=20, dev=0.1)
# pc = pc.grid()
# pc.write("../../data/file26new.pcd")

# print(pc.size())
# for i in range(pc.size()):
#     print(pc.get(i))
